### ISSUE TITLE
Kubernetes Basic Concepts

### ISSUE THREAD

**Issue by codeanit** [codeanit/til#19]
![Modernization Of Deployment Processes](https://d33wubrfki0l68.cloudfront.net/26a177ede4d7b032362289c6fccd448fc4a91174/eb693/images/docs/container_evolution.svg)

_`Write programs that do one thing and do it well - The Unix Philosophy`_

### Brief History
> Software systems runs country and has now become part of the society. Continuous updates, upgrades and maintenance has been a necessity without which many life support system fails. The Unix philosophy helped address these needs, and the applications stated to be more distributed with the advent of The Cloud.  
Traditionally, organizations ran one or many applications on physical one physical servers. There were many issues to manage the system and the applications but the main drawback was that it did not scale well. To scale the system and for optimum use of the resources, Virtual Machines(VM) were used. It also provided better security as applications can now be isolated as needed. The nature of application started to be complex. Today we have Containerized applications that abstracted and enhanced the entire functionalities of VM's, making it exceptionally easy to maintain distributed systems.    

![Kubernetes](https://user-images.githubusercontent.com/414141/85718992-c32aab00-b70e-11ea-9f65-2c1310c1c0a6.png)

The following is a summary of the key API objects of Kubernetes:
- Clusters: Pool of compute, storage, and network resources.
- Nodes: Host machines running within the cluster.
- Namespaces: Logical partitions of a cluster.
- Pods: Units of deployment.
- Labels and Selectors: Key-Value pairs for identification and service discovery.
- Services: Collection of Pods belonging to the same application.
- Replica Set: Ensures availability and scalability.
- Deployment: Manages application lifecycle.

Container: A container is a standard unit of software that packages up code and all its dependencies so the application runs quickly and reliably from one computing environment to another. 

Kubernetes: A container orchestration tool. It uses containers to run isolated, packaged applications across its cluster nodes. 

Cluster: A group of interconnected `Node`. Cluster's state is defined `Kubernetes Objects`. 
Cluster's desired state includes what applications or other workloads to run, what container images they use, the number of replicas, what network and disk resources to make available.

Node: Nodes are physical or virtual machines. Every cluster must have at least one `Master` node which controls cluster, and one or many `Worker` nodes which runs containers.

Kubernetes Objects: These are persistent entities in Kubernetes that defines everything in Kubernetes does. Objects are stored in `etcd`as a key-value pair.
Pod, Service, Volume and Namespace are the basic Kubernetes Objects which determines the deployed containerized applications and workloads, their associated network and disk resources. 
Deployment, DaemonSet, StatefulSet, ReplicaSet and Job are the other Objects which provide additional functionality and convenience features. These are build upon basic objects. 

Kubernetes API: A HTTP API to create the Kubernetes Objects and the communication medium for the end users, different parts of your cluster, and external components with one another.

Controllers: A control loop is a non-terminating loop that regulates the state of a system. Controllers are control loops that watch the state of your cluster, then make or request changes where needed. Each controller tries to move the current cluster state closer to the desired state. `kube-controller-manager` runs controller processes.
   
Kubernetes Control Plane: A collection Controllers. It consists of Master Node.

Pod: It represents a group of one or more containers running together on your cluster. It’s the possible smallest object deployed in the cluster. 

Volume: A volume is simply a directory of data; it lives within a pod and can be accessed by any container running in that pod.
A volume is the abstraction that lets Kubernetes deal with the ephemeral nature of containers; when a container is retired, the volume (and its data) continues to exist within the pod, still accessible to other containers. It exists as long as its pod exists; once the latter “dies,” so does the volume and its data.

Persistent Volume: Persistent volumes deal with the issue of storage that needs to exist outside of the lifetime of any particular container or application, whereas general volumes deal with compute. This becomes
particularly important when you’re discussing stateful applications like databases.

Master Node: It runs three critical processes, i.e. kube-apiserver, kube-controller-manager and kube-scheduler. The master can also be replicated for availability and redundancy.
 
Worker Node: These are the machines that run containerized applications. The worker node(s) host the Pods that are the components of the application workload. Each individual Worker Node in the cluster runs two processes: kubelet and kube-proxy.

`kube-apiserver` 55t s the implementation of the Kubernetes API.
`kubectl` is a command-line interface tool to interact with the Kubernetes API.
`kube-scheduler` watches for newly created Pods with no assigned node, and selects a node for them to run on.
`kubelet` Communicates with the Master.
`kube-proxy` A network proxy which reflects Kubernetes networking services on each node.

> Types Of Controllers
`Node Controller`: Responsible for noticing and responding when nodes go down.
`Replication Controller`: Responsible for maintaining the correct number of pods for every replication controller object in the system.
`Endpoints Controller`: Populates the Endpoints object (that is, joins Services & Pods).
`Service Account & Token Controllers`: Create default accounts and API access tokens for new namespaces

Logically, each controller is a separate process, but to reduce complexity, they are all compiled into a single binary and run in a single process. 

> References
- https://kubernetes.io/docs/concepts
- https://www.digitalocean.com/community/curriculums/kubernetes-for-full-stack-developers

**Comment by codeanit** [comment]
Could not work on this issue today.

Found two interesting links to read:

- https://learnk8s.io/real-time-dashboard
- https://enterprisersproject.com/article/2020/4/kubernetes-architecture-beginners

**Comment by codeanit** [comment]
- [ ] Found a [cheatsheet](https://opensource.com/sites/default/files/gated-content/cheat_sheet_kubectl.pdf)

**Comment by codeanit** [comment]
https://enterprisersproject.com/article/2020/4/kubernetes-architecture-beginners
https://kubernetes.io/docs/concepts/overview/components
https://enterprisersproject.com/article/2020/4/kubernetes-everything-you-need-know
https://www.redhat.com/en/events/webinar/kubernetes-101%E2%80%94-introduction-containers-kubernetes-and-openshift-red-hat-training?intcmp=701f2000000tjyaAAA
https://enterprisersproject.com/tags/containers 


https://opensource.com/sites/default/files/gated-content/getting_started_with_kubernetes.pdf
https://enterprisersproject.com/sites/default/files/kubernetes_glossary.pdf
https://opensource.com/sites/default/files/gated-content/cheat_sheet_kubectl.pdf


https://enterprisersproject.com/article/2019/4/minikube-how-to-use
https://istio.io/docs/setup/platform-setup/minikube
https://operatorhub.io | https://github.com/operator-framework


Need to go through these resources again.

**Comment by codeanit** [comment]
Make a note based on https://kubernetes.io/docs/concepts/overview/components and other articles to close the section and create tasks on other specific concepts to understand them in depth.

**Comment by codeanit** [comment]
- [ ] https://enterprisersproject.com/article/2020/5/kubernetes-5-open-source-projects-improve

**Comment by codeanit** [comment]
https://enterprisersproject.com/article/2020/5/kubernetes-migration-4-secrets
https://enterprisersproject.com/article/2020/5/kubernetes-5-open-source-projects-improve
https://enterprisersproject.com/article/2020/5/kubernetes-deployments-6-security-best-practices

**Comment by codeanit** [comment]
- [ ] TODO: Examples of Deployment and Service templates

**Comment by anitsh** [comment]
# ReplicationController
A ReplicationController is responsible for managing the pod lifecycle and ensuring that the specified number of pods required are running at any given time.

A ReplicationController is responsible for managing the pod lifecycle and ensuring that the specified number of pods required are running at any given time. On the other hand, it is not responsible for the advanced cluster capabilities like performing auto-scaling, readiness and liveliness probes, and other advanced replication capabilities. Other components within the Kubernetes cluster better perform those capabilities.

In short, the ReplicationController has limited responsibility and gets typically used for specific implementations that do not require complex logic to attain certain requirements (for example, ensuring that the desired number of pods always matches the specified number). If there are more than the desired number, the ReplicationController removes the excess ones and ensures the same number of pods exist even in the event of node failure or pod termination.

https://opensource.com/article/21/11/kubernetes-replicationcontroller
